packages/gossip/manager.go

```diff
@@ -200,65 +224,76 @@ func (m *Manager) getNeighborsByID(ids []identity.ID) []*Neighbor {
 	return result
 }
 
-func (m *Manager) send(b []byte, to ...identity.ID) {
+func (m *Manager) send(packet *pb.Packet, to ...identity.ID) {
 	neighbors := m.getNeighborsByID(to)
 	if len(neighbors) == 0 {
 		neighbors = m.AllNeighbors()
 	}
 
 	for _, nbr := range neighbors {
-		if _, err := nbr.Write(b); err != nil {
+		if err := nbr.write(packet); err != nil {
 			m.log.Warnw("send error", "peer-id", nbr.ID(), "err", err)
 		}
 	}
 }
 
 func (m *Manager) addNeighbor(ctx context.Context, p *peer.Peer, group NeighborsGroup,
-	connectorFunc func(context.Context, *peer.Peer, ...server.ConnectPeerOption) (net.Conn, error),
-	connectOpts []server.ConnectPeerOption,
+	connectorFunc func(context.Context, *peer.Peer, []ConnectPeerOption) (network.Stream, error),
+	connectOpts []ConnectPeerOption,
 ) error {
 	if p.ID() == m.local.ID() {
-		return ErrLoopbackNeighbor
+		return errors.WithStack(ErrLoopbackNeighbor)
 	}
-	m.serverMutex.RLock()
-	defer m.serverMutex.RUnlock()
-	if m.server == nil {
+	m.stopMutex.RLock()
+	defer m.stopMutex.RUnlock()
+	if m.isStopped {
 		return ErrNotRunning
 	}
-	if m.neighborExists(p.ID()) {
-		m.neighborsEvents[group].ConnectionFailed.Trigger(p, ErrDuplicateNeighbor)
-		return ErrDuplicateNeighbor
+	exists := m.neighborExists(p)
+	if exists {
+		return errors.WithStack(ErrDuplicateNeighbor)
 	}
 
-	conn, err := connectorFunc(ctx, p, connectOpts...)
+	stream, err := connectorFunc(ctx, p, connectOpts)
 	if err != nil {
-		m.neighborsEvents[group].ConnectionFailed.Trigger(p, err)
-		return err
+		return errors.WithStack(err)
 	}
 
 	// create and add the neighbor
-	nbr := NewNeighbor(p, group, conn, m.log)
+	nbr := NewNeighbor(p, group, stream, m.log)
 	if err := m.setNeighbor(nbr); err != nil {
-		_ = conn.Close()
-		m.neighborsEvents[group].ConnectionFailed.Trigger(p, err)
+		if resetErr := stream.Close(); resetErr != nil {
+			err = errors.CombineErrors(err, resetErr)
+		}
 		return errors.WithStack(err)
 	}
+	nbr.disconnected.Attach(events.NewClosure(func() {
+		m.deleteNeighbor(nbr.ID())
+		go m.NeighborsEvents(nbr.Group).NeighborRemoved.Trigger(nbr)
+	}))
+	nbr.packetReceived.Attach(events.NewClosure(func(packet *pb.Packet) {
+		if err := m.handlePacket(packet, nbr); err != nil {
```

Comment:
The copy was done on `[]byte` because the gossip manager was operating and mutating bytes, it would mess up the BuffConn reading logic. In the new implementation we decode bytes into protobuf structs as soon as possible and don't operate on `[]byte`, so there is no point copying.